nbbuild/antsrc/org/netbeans/nbbuild/ModuleListParser.java

```diff
@@ -339,35 +353,24 @@ private static boolean scanPossibleProject(File dir, Map<String,Entry> entries,
         faketask.setName("module.jar");
         faketask.setValue(fakeproj.replaceProperties("${module.jar.dir}/${module.jar.basename}"));
         faketask.execute();
+        String id = null;
         switch (moduleType) {
         case NB_ORG:
             assert path != null;
             // Find the associated cluster.
             // first try direct mapping in nbbuild/netbeans/moduleCluster.properties
-            String clusterDir = (String) properties.get(path + ".dir");
-            if (clusterDir != null) {
-                clusterDir = clusterDir.substring(clusterDir.lastIndexOf('/') + 1);
-            } else {
-                // not found, try indirect nbbuild/cluster.properties
-                for (Map.Entry<String,Object> entry : properties.entrySet()) {
-                    String val = (String) entry.getValue();
-                    String[] modules = val.split(", *");
-                    if (Arrays.asList(modules).contains(path)) {
-                        String key = entry.getKey();
-                        clusterDir = (String) properties.get(key + ".dir");
-                        if (clusterDir != null) {
-                            faketask.setName("cluster.dir");
-                            faketask.setValue(clusterDir);
-                            faketask.execute();
-                            break;
-                        }
-                    }
+            String[] clusterDir = { (String) properties.get(path + ".dir") };
+            if (clusterDir[0] != null) {
+                final String subStr = clusterDir[0].substring(clusterDir[0].lastIndexOf('/') + 1);
+                if (path.startsWith(subStr + "/")) {
+                    id = path.substring(subStr.length() + 1);
                 }
-                if (clusterDir == null)
-                    clusterDir = "extra";   // fallback
+                clusterDir[0] = subStr;
+            } else {
+                id = SetCluster.findClusterAndId("cluster.dir", clusterDir, properties, path, faketask, "extra");
```

Comment:
The same algorithm was written twice before - trying to move it into single `SetCluster.findClusterAndId` place.